File: src-web/src/protocol.ts
/**
 * Protocol Types
 *
 * WebSocket message types mirroring Rust protocol.rs EXACTLY
 *
 * Design:
 * - Request/Response pairs co-located in namespaces (mirrors Rust modules)
 * - ClientMessage and ServerMessage discriminated unions
 * - Simple, clear structure with compile-time type safety
 */

import { AXNode, AXValue } from "./axio.ts";

// ============================================================================
// Window Types
// ============================================================================

export interface Window {
  id: string;
  title: string;
  app_name: string;
  x: number;
  y: number;
  w: number;
  h: number;
  focused: boolean;
  root?: AXNode; // Accessibility tree root (populated by backend when available)
}

// ============================================================================
// Request/Response Pair Namespaces
// ============================================================================

/** Get children of an accessibility element */
export namespace GetChildren {
  export interface Request {
    element_id: string;
    max_depth?: number;
    max_children_per_level?: number;
  }

  export interface Response {
    success: boolean;
    children?: AXNode[];
    error?: string;
  }
}

/** Write text to an accessibility element */
export namespace WriteToElement {
  export interface Request {
    element_id: string;
    text: string;
  }

  export interface Response {
    success: boolean;
    error?: string;
  }
}

/** Click an accessibility element */
export namespace ClickElement {
  export interface Request {
    element_id: string;
  }

  export interface Response {
    success: boolean;
    error?: string;
  }
}

/** Enable/disable click-through on overlay window */
export namespace SetClickthrough {
  export interface Request {
    enabled: boolean;
  }

  export interface Response {
    success: boolean;
    enabled: boolean;
    error?: string;
  }
}

/** Start watching an element for changes */
export namespace WatchNode {
  export interface Request {
    element_id: string;
    node_id: string;
  }

  export interface Response {
    success: boolean;
    node_id: string;
    error?: string;
  }
}

/** Stop watching an element */
export namespace UnwatchNode {
  export interface Request {
    element_id: string;
  }

  export interface Response {
    success: boolean;
  }
}

/** Get accessibility element at screen position */
export namespace GetElementAtPosition {
  export interface Request {
    x: number;
    y: number;
  }

  export interface Response {
    success: boolean;
    element?: AXNode;
    error?: string;
  }
}

// ============================================================================
// Client -> Server Messages
// ============================================================================

export type ClientMessage =
  | ({ type: "get_children" } & GetChildren.Request)
  | ({ type: "write_to_element" } & WriteToElement.Request)
  | ({ type: "click_element" } & ClickElement.Request)
  | ({ type: "set_clickthrough" } & SetClickthrough.Request)
  | ({ type: "watch_node" } & WatchNode.Request)
  | ({ type: "unwatch_node" } & UnwatchNode.Request)
  | ({ type: "get_element_at_position" } & GetElementAtPosition.Request);

// ============================================================================
// Server -> Client Messages
// ============================================================================

export type ServerMessage =
  // Push Events
  | { type: "window_update"; windows: Window[] }
  | { type: "window_root_update"; window_id: string; root: AXNode }
  | { type: "mouse_position"; x: number; y: number }
  | { type: "element_update"; update: ElementUpdate }
  // Response Messages
  | ({ type: "get_children_response" } & GetChildren.Response)
  | ({ type: "write_to_element_response" } & WriteToElement.Response)
  | ({ type: "click_element_response" } & ClickElement.Response)
  | ({ type: "set_clickthrough_response" } & SetClickthrough.Response)
  | ({ type: "watch_node_response" } & WatchNode.Response)
  | ({ type: "unwatch_node_response" } & UnwatchNode.Response)
  | ({
      type: "get_element_at_position_response";
    } & GetElementAtPosition.Response);

// ============================================================================
// Element Update Types
// ============================================================================

export type ElementUpdate =
  | { update_type: "ValueChanged"; element_id: string; value: AXValue }
  | { update_type: "LabelChanged"; element_id: string; label: string }
  | { update_type: "ElementDestroyed"; element_id: string };

// ============================================================================
// Optional: Helper Constructors (add as needed)
// ============================================================================

export namespace GetChildren {
  export function success(children: AXNode[]): Response {
    return { success: true, children };
  }

  export function error(error: string): Response {
    return { success: false, error };
  }
}

export namespace WriteToElement {
  export function success(): Response {
    return { success: true };
  }

  export function error(error: string): Response {
    return { success: false, error };
  }
}

export namespace ClickElement {
  export function success(): Response {
    return { success: true };
  }

  export function error(error: string): Response {
    return { success: false, error };
  }
}

export namespace SetClickthrough {
  export function success(enabled: boolean): Response {
    return { success: true, enabled };
  }

  export function error(enabled: boolean, error: string): Response {
    return { success: false, enabled, error };
  }
}

export namespace WatchNode {
  export function success(node_id: string): Response {
    return { success: true, node_id };
  }

  export function error(node_id: string, error: string): Response {
    return { success: false, node_id, error };
  }
}

export namespace UnwatchNode {
  export function success(): Response {
    return { success: true };
  }
}

export namespace GetElementAtPosition {
  export function success(element: AXNode): Response {
    return { success: true, element };
  }

  export function error(error: string): Response {
    return { success: false, error };
  }
}

File: src-tauri/src/protocol.rs
/**
 * Protocol Types
 *
 * Complete WebSocket message protocol for client-server communication.
 * All communication types are defined here for clarity and type safety.
 *
 * Design:
 * - Request/Response pairs co-located in modules
 * - ClientMessage and ServerMessage enums for serialization
 * - Simple, clear structure with compile-time type safety
 */
use serde::{Deserialize, Serialize};

use crate::axio::{AXNode, AXValue};
use crate::windows::WindowInfo;

// ============================================================================
// Request/Response Pair Modules
// ============================================================================

/// Get children of an accessibility element
pub mod get_children {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Request {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub element_id: String,
        #[serde(default = "super::default_max_depth")]
        pub max_depth: usize,
        #[serde(default = "super::default_max_children")]
        pub max_children_per_level: usize,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Response {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub success: bool,
        pub children: Option<Vec<AXNode>>,
        pub error: Option<String>,
    }
}

/// Write text to an accessibility element
pub mod write_to_element {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Request {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub element_id: String,
        pub text: String,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Response {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub success: bool,
        pub error: Option<String>,
    }
}

/// Click an accessibility element
pub mod click_element {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Request {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub element_id: String,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Response {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub success: bool,
        pub error: Option<String>,
    }
}

/// Enable/disable click-through on overlay window
pub mod set_clickthrough {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Request {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub enabled: bool,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Response {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub success: bool,
        pub enabled: bool,
        pub error: Option<String>,
    }
}

/// Start watching an element for changes
pub mod watch_node {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Request {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub element_id: String,
        pub node_id: String,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Response {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub success: bool,
        pub node_id: String,
        pub error: Option<String>,
    }
}

/// Stop watching an element
pub mod unwatch_node {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Request {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub element_id: String,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Response {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub success: bool,
    }
}

/// Get accessibility element at screen position
pub mod get_element_at_position {
    use super::*;

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Request {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub x: f64,
        pub y: f64,
    }

    #[derive(Debug, Clone, Serialize, Deserialize)]
    pub struct Response {
        #[serde(skip_serializing_if = "Option::is_none")]
        pub request_id: Option<String>,
        pub success: bool,
        pub element: Option<AXNode>,
        pub error: Option<String>,
    }
}

// ============================================================================
// Client -> Server Messages
// ============================================================================

/// All messages that can be sent from client to server
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMessage {
    GetChildren(get_children::Request),
    WriteToElement(write_to_element::Request),
    ClickElement(click_element::Request),
    SetClickthrough(set_clickthrough::Request),
    WatchNode(watch_node::Request),
    UnwatchNode(unwatch_node::Request),
    GetElementAtPosition(get_element_at_position::Request),
}

// ============================================================================
// Server -> Client Messages
// ============================================================================

/// All messages that can be sent from server to client
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    // Push Events (not request/response)
    WindowUpdate { windows: Vec<WindowInfo> },
    WindowRootUpdate { window_id: String, root: AXNode },
    MousePosition { x: f64, y: f64 },
    ElementUpdate { update: ElementUpdate },

    // Response Messages (paired with requests above)
    GetChildrenResponse(get_children::Response),
    WriteToElementResponse(write_to_element::Response),
    ClickElementResponse(click_element::Response),
    SetClickthroughResponse(set_clickthrough::Response),
    WatchNodeResponse(watch_node::Response),
    UnwatchNodeResponse(unwatch_node::Response),
    GetElementAtPositionResponse(get_element_at_position::Response),
}

// ============================================================================
// Element Update Types (Server Push Events)
// ============================================================================

/// Update events for accessibility elements
/// Moved from axio.rs - this is part of the protocol, not core AXIO types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "update_type", rename_all = "PascalCase")]
pub enum ElementUpdate {
    ValueChanged { element_id: String, value: AXValue },
    LabelChanged { element_id: String, label: String },
    ElementDestroyed { element_id: String },
}

// ============================================================================
// Helper Functions
// ============================================================================

fn default_max_depth() -> usize {
    50
}

fn default_max_children() -> usize {
    2000
}
